### PR TITLE
Factor lock_table out into a new module.

### PR DESCRIPTION
**Pull request by mystenmark**
I'll be using this for the WAL work.

Happy to make a new crate for this kind of stuff if we don't think it belongs in sui-storage.

### CODE REVIEW

**Comment by bmwill** [comment]
I'm still not quiet sure I understand why we need mutexes for this. If we have two txns with the exact same digest enter this function we only want one of them to succeed right? Why can't we have one of them bail out of the function instead of waiting for the other to finish?

File: crates/sui-storage/src/mutex_table.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            lock_table: (0..size)
+                .into_iter()
+                .map(|_| tokio::sync::Mutex::new(()))
+                .collect(),
+            _k: std::marker::PhantomData {},
+        }
+    }
+
+    fn get_lock_idx(&self, key: &K) -> usize {
+        let mut hasher = Blake2s::new();
```
Comment (gdanezis) [inline crates/sui-storage/src/mutex_table.rs:37]: I would honestly not hash using a secure hash function here, this is an expensive operation.

File: crates/sui-storage/src/mutex_table.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        hasher.update(self.hash_secret);
+        hasher.update(key);
+        let digest = hasher.finalize();
+        usize::from_le_bytes(digest[0..8].try_into().unwrap()) % self.lock_table.len()
+    }
+
+    pub async fn acquire_locks<'a>(
+        &'a self,
+        objects: &[K],
```
Comment (gdanezis) [inline crates/sui-storage/src/mutex_table.rs:46]: should we get an iterator to `&K` instead?

File: crates/sui-storage/src/mutex_table.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    ) -> Vec<tokio::sync::MutexGuard<'a, ()>> {
```
Comment (gdanezis) [inline crates/sui-storage/src/mutex_table.rs:47]: Lets make a special opaque type for `Vec<tokio::sync::MutexGuard<'a, ()>>`?

**Review by gdanezis — APPROVED** [review]
Nice!

**Comment by gdanezis** [comment]
> I'm still not quiet sure I understand why we need mutexes for this. If we have two txns with the exact same digest enter this function we only want one of them to succeed right? Why can't we have one of them bail out of the function instead of waiting for the other to finish?

One day we should extend this class to wrap it around the digests we are reading / writing and not allow access to them unless you hold a valid guard. Until then, this is fine although more in line with what we used to do in the dark days of C/C++ rather than in Rust.

**Comment by mystenmark** [comment]
> I'm still not quiet sure I understand why we need mutexes for this.

in the current use case, we need to prevent multiple txes (including duplicate concurrent txes) from operating on the same locks concurrently.

later, i may relax this to only prevent the same TX from running twice concurrently, but this is still an excellent and extremely fast way to do that.

**Comment by bmwill** [comment]
> in the current use case, we need to prevent multiple txes (including duplicate concurrent txes) from operating on the same locks concurrently.
> 
> later, i may relax this to only prevent the same TX from running twice concurrently, but this is still an excellent and extremely fast way to do that.

Look i still think this is a terrible idea and that the problem could be restructured to enable a cleaner solution.

Also i've asked this question numerous times and no one has yet to give me an answer:

> Why can't we have one of them bail out of the function instead of waiting for the other to finish?

The function that has this critical section does not have a return type that requires the caller getting any information. Instead we could simply bail since if some other txn is already trying to write to the store, with the exact same digest, we know that the second one will do nothing anyway.

```
    async fn batch_update_objects<BackingPackageStore>(
        &self,
        mut write_batch: DBBatch,
        temporary_store: AuthorityTemporaryStore<BackingPackageStore>,
        transaction_digest: TransactionDigest,
        seq_opt: Option<TxSequenceNumber>,
    ) -> Result<(), SuiError> {
```

File: crates/sui-storage/src/mutex_table.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-storage/src/mutex_table.rs:37]: switched to using the DefaultHasher used by hash_map

**Comment by mystenmark** [comment]
> Also i've asked this question numerous times and no one has yet to give me an answer:
> 
> > Why can't we have one of them bail out of the function instead of waiting for the other to finish?


> The function that has this critical section does not have a return type that requires the caller getting any information. Instead we could simply bail since if some other txn is already trying to write to the store, with the exact same digest, we know that the second one will do nothing anyway

I'm sorry if this hasn't been answered clearly before, let me make one more try:

We _do_ bail if there are concurrent updates, since the loser of the race will find that the locks have been deleted. The critical section **is required into order to do that check correctly**.
 
Also, that function (`batch_update_objects`) probably won't be the place where the locking will be done after I am finished with this work.

Right now I am trying to move some code around, and I can't also re-think the safety of the critical section that depends on this behavior at the same time.

**Comment by bmwill** [comment]
> We do bail if there are concurrent updates, since the loser of the race will find that the locks have been deleted. The critical section is required into order to do that check correctly.

These "locks" are the actual object locks, not mutexes, correct? In that case is there a reason why we couldn't push such a check earlier by checking a Hashset of current digests currently writing objects? (Such a hashset could be implemented to be fast despite needing a mutex over the structure via sharding)

> Right now I am trying to move some code around, and I can't also re-think the safety of the critical section that depends on this behavior at the same time.

I can understand that. I just really really don't want this paradigm to be widely used and i worry that making it an easy to instantiate component will mean it will be used more broadly when instead we should be eliminating it entirely.

**Comment by mystenmark** [comment]
> These "locks" are the actual object locks, not mutexes, correct? In that case is there a reason why we couldn't push such a check earlier by checking a Hashset of current digests currently writing objects? (Such a hashset could be implemented to be fast despite needing a mutex over the structure via sharding)

like i said, i may be moving the check earlier, so I'm happy to discuss that when it comes up.

re HashSets - what do you see as the advantage over the current approach? It eliminates a rare sub-optimal case (false exclusion due to hash collision) by making every case (very slightly) slower. I doubt it would have any measurable perf impact in either direction.

**Comment by velvia** [comment]
> Look i still think this is a terrible idea and that the problem could be restructured to enable a cleaner solution.

> Also i've asked this question numerous times and no one has yet to give me an answer:

> Why can't we have one of them bail out of the function instead of waiting for the other to finish?
The function that has this critical section does not have a return type that requires the caller getting any information. Instead we could simply bail since if some other txn is already trying to write to the store, with the exact same digest, we know that the second one will do nothing anyway.

Well, this has been merged and it's fine, it's a really small change.  

However, I do think that the ball of wax that is the current AuthorityStore and the locking/critical section stuff, is definitely difficult to understand in detail, and not easy for newcomers to jump in.  I think we are definitely overdue for a rethink or at least examination of how it works, at least so that others can participate more.

**Pull request merged**